Fake data Issue #354

VikramsDataScience opened the pull request:
# Pull Request
Issue #354

## Description
- Changed the `.pre-commit-config.yaml` file to use python 3.11.5.
- I also noticed that the `requirements.txt` didn't have the `pre-commit` library listed. So I modified it to include this library. From what I can tell it conforms to their best practice (https://pre-commit.com/#install). I hope that's okay? Please let me know if you want me to revert, though :).
- Made the changes to all 6 routes of the API to incorporate an `is_fake()` environment in the `gsp.py` module. I've attempted to make the changes to mimic the [module](https://github.com/openclimatefix/pv-site-api/blob/main/pv_site_api/main.py#L199). @peterdudfield could you let me know if this is along the lines of what you're after, please? And, based on your feedback, I can make the necessary changes in the follow up commits :).
- For the fake forecast generation, I've used the `make_fake_forecast()` function from the `nowcasting_datamodel.fake` module. But, there's also the 'plural' `make_fake_forecasts()` function available, that seems to call the same `make_fake_forecast()` function, but for multiple grid supply point IDs. Could you also advise if this is correct, please?

Fixes #

## How Has This Been Tested?
- For testing purposes I've slightly modified the relative imports in `auth_utils`, `cache.py`, `database.py`, `pydantic_models`, and `utils.py` modules to allow testing the fake data in all the API routes of the `gsp.py` module. If all goes well and you're happy with the changes in the PR, I can revert the changes in the relative imports of the other modules prior to approval/merge.
- I've also set the sys path so that the `nowcasting_datamodel` repo can be used for relative imports
- From what I can tell, the changes seem to run on my local machine with the exception of the following warnings - which seem unrelated to the fake data changes but due to some updates that may need to be made to the `pydantic` library? I'm not too sure, though:
![image](https://github.com/user-attachments/assets/b02acb2b-a8b7-4c14-a2a5-173e845c5279)
- [x] Yes

_If your changes affect data processing, have you plotted any changes? i.e. have you done a quick sanity check?_
**I haven't plotted any of the changes, other than running the test with the screenshot above. But, please advise if there's anything specific you'd like to see?**
- [ ] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation **Not sure if any documentation changes are needed for this change. But, I've added a docstring to the is_fake() function.**
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

peterdudfield commented:
Thanks so much @VikramsDataScience for this. Really appreciate you doing this. 

I wonder if there's a few more things you could do? 

1. Could you check that when you run things locally, could you manually check that the routes your modified work? And they look like the return object? You might find some are empty?

2. Would you be able to add a test or two, to check they work? This means in the future we are really onfident about this working

3. Bonus: If you fancy running the front end using https://github.com/openclimatefix/quartz-frontend/tree/main/apps/nowcasting-app and seeing if the app looks like it would in live. You can access the live service at app.quartz.solar

Thanks once again for doing all this work, really appreciate that your doing this. No pressure, to do this extra work, just say.

VikramsDataScience commented:
Hey @peterdudfield. Sure thing. I'm happy to do it :)!

If its cool with you, I'll look to get this done sometime next week? This is only because it's a public holiday in Melbourne tomorrow, and I fancied having a 3 day weekend :).

peterdudfield commented:
Thanks

No worries, do this when it suits you

VikramsDataScience commented:
Hey @peterdudfield. I just wanted to provide a quick update: 
- I'm just now looking to test the changes. I've initially had quite a lot of trouble with Docker (on my local). I've got some issues with WSL2 setting up Docker on Desktop. Seems like it cannot set up the Linux Engine locally for some reason. So I haven't been able to test this with Docker. It took me quite a few hours of messing around (including with playing with my BIOS settings) to realise that it wouldn't work!
- So, I've instead opted to set up the PostGres DB locally for which you guys have already created data models in the `nowcasting_datamodel`. I've just been able to get it running on my local, I'll try to get some tests running. Please bear with me in the meantime! I'll come back with more questions if I hit upon any blockers :).

peterdudfield commented:
No problem 

Please do share any learnings, just so we can help others too

VikramsDataScience commented:
Hey @peterdudfield. So sorry its taking me so long to update this PR since last week! It's been a bit of a distracting and chaotic period lately.

I've tried to write a test case to test the changes for the fake forecast, and it may be generating forecasts, but I'm not really sure and wanted to get your thoughts, please? 

I've made a number of changes to various modules, in my local environment, so I didn't want to include them as commits, as they're only done to test the changes, and won't be part of the eventual merge into production. Instead I wanted to just provide some snippets of the changes, and which modules they affect, and you could let me know if I'm on the right track? And, once we're settled on what is correct, I can then include everything into a commit?

- With the test cases, I've written one possible test case to begin with that's attempted to mimic some of the other similar test cases in the `test_gsp.py` module. This one was intended to test the `get_forecasts_for_a_specific_gsp()` function (/{gsp_id}/forecast route) from the `gsp.py` module, just to kick things off before looking at the other routes/endpoints:
```
@freeze_time("2024-01-01")
def test_forecasts_for_specific_gsps(db_session, api_client, gsp_id=1):
    """Test if route /v0/solar/GB/gsp/{gsp_id}/forecast works"""

    # Using some dummy values
    forecast_value = [
        ForecastValueSQL(
            target_time=datetime(2024, 2, 1, tzinfo=timezone.utc),
            expected_power_generation_megawatts=50.0,
            adjust_mw=0.2,
            properties={},  
            forecast_id=1,  
            created_utc=datetime(2024, 2, 1, tzinfo=timezone.utc)
        )
    ]

    forecasts = make_fake_forecast(
        gsp_id=gsp_id,
        session=db_session,
        forecast_values=forecast_value,
        add_latest=True, 
        model_name="fake_model"
        )
    db_session.add(forecasts)
    db_session.commit()

    forecast_from_db = db_session.query(ForecastSQL).filter_by(id=forecasts.id).first()

    # Print the values in the DB
    print(f"Stored Forecast ID: {forecast_from_db.id}")
    print(f"Forecast creation time: {forecast_from_db.forecast_creation_time}")
    for value in forecast_from_db.forecast_values:
        print(f"Stored Target time: {value.target_time},\n Stored Expected power: {value.expected_power_generation_megawatts}")

    app.dependency_overrides[get_session] = lambda: db_session

    response = api_client.get(f"/v0/solar/GB/gsp/{gsp_id}/forecast")

    assert response.status_code == 200

    return [ForecastSQL(**f) for f in response.json()]
```
This test case returns the following output. So, it looks like I'm getting a connection to the DB for the route/endpoint and that allows the assert in the test case to pass, and I'm seeing something populating in the 'expected_power_generation_megawatts' value (I've highlighted it on the screenshot). The other outputs seem to be print statements coming from the `get_forecasts_for_a_specific_gsp()` function in the `gsp.py` module. I'm really not sure if this correct, though. Can you advise, please?:
![image](https://github.com/user-attachments/assets/4c92989c-2cf8-4aa4-bf3c-08e6c58768e4)

VikramsDataScience commented:
Hey @peterdudfield. With regards to the changes I was referencing in yesterday's lengthy comment, I've put them into a commit and tested it. Could you kindly review, and let me know, please?

VikramsDataScience commented:
- Hey @peterdudfield. I've made another push to the PR. When you return, could review the changes and let me know, please?
- Updated the `gsp.py` module to recursively generate "gsp_ids" for the `get_all_available_forecasts()` and `get_truths_for_all_gsps()` functions. Both these functions seem to make respective calls using the `get_forecasts_from_database()` and `get_truth_values_for_all_gsps_from_database()` that returns 0 forecasts. Please correct me if I'm wrong, but I think this should expected, since the `make_fake_forecasts()` call in the test case I've written should take over that duty, and override the need to connect to the DB?
- From what I can understand of the DB data model, it seems that the forecasts are stored in `forecast_values.expected_power_generation_megawatts`? If you could please confirm this, as I'm not 100% sure if that's correct, and I can change the test cases if necessary?

peterdudfield commented on src/tests/test_gsp.py:
```diff
@@ -418,3 +421,163 @@ def test_read_truths_for_all_gsp_compact(db_session, api_client):
+
+
+@freeze_time("2024-01-01")
+def test_forecasts_for_specific_gsps(db_session, api_client, gsp_id=1):
```
isnt this covered in line 9

peterdudfield commented on src/tests/test_gsp.py:
```diff
@@ -418,3 +421,163 @@ def test_read_truths_for_all_gsp_compact(db_session, api_client):
+
+
+@freeze_time("2024-01-01")
+def test_get_truths_for_all_gsps(db_session, api_client):
```
isnt this tested in 366?

peterdudfield commented on src/tests/test_gsp.py:
```diff
@@ -418,3 +421,163 @@ def test_read_truths_for_all_gsp_compact(db_session, api_client):
+
+@freeze_time("2024-01-01")
+def test_get_all_available_forecasts(db_session, api_client):
+    """Test if route /v0/solar/GB/gsp/forecast/all/ works"""
```
is this a copy of line 99?

peterdudfield commented on src/tests/test_gsp.py:
```diff
@@ -418,3 +421,163 @@ def test_read_truths_for_all_gsp_compact(db_session, api_client):
+
+@freeze_time("2024-01-01")
+def test_get_truths_for_a_specific_gsp(db_session, api_client, gsp_id=1):
+    """Test is route /v0/solar/GB/gsp/{gsp_id}/pvlive works
```
copy of 257?

VikramsDataScience commented:
Hey @peterdudfield. No worries at all! I think you were on leave?

Apologies for the duplicated test cases. When you were asking about writing some tests, I didn't think to look at the existing test cases that were already there! I've now removed the duplicated test cases, and pushed the changes to the PR. 

Please let me know if there's anything else?

peterdudfield commented:
> Hey @peterdudfield. No worries at all! I think you were on leave?
> 
> Apologies for the duplicated test cases. When you were asking about writing some tests, I didn't think to look at the existing test cases that were already there! I've now removed the duplicated test cases, and pushed the changes to the PR.
> 
> Please let me know if there's anything else?

yea, I was away for a week. 

Sorry I should have said this yesterday, yea, test would be good, but ones that test `is_fake` would be really useful

VikramsDataScience commented:
All good. Sorry for the delayed response, but I've had my week hijacked with some other important priorities. Could I look into developing some new test cases sometime next week, please?

peterdudfield commented:
Of course,no problem. Thsnk you for your help

VikramsDataScience commented:
Of course! A pleasure, as always, @peterdudfield!

VikramsDataScience commented:
Hey @peterdudfield. So sorry for my tardy updates with this PR! It's been a very distracting period for me, and has been a bit difficult to manage some of my competing priorities.

I've written a new test case for the fake environment that has passed the checks on my local machine, and is generating forecasts for one of the DB routes/endpoints (I thought I'd start with the specific GSP ID test to begin). It's generating quite a few 0.0 forecasts, but happily no negative values:
![image](https://github.com/user-attachments/assets/6cf8237d-df5a-4af3-87e4-551e5291c584)
Could you review the test case and let me know if this is okay with you, please? And, if all is good with you, I can proceed to write very similar test cases for the remaining routes/endpoints :).

VikramsDataScience commented:
> No worries. This looks great.
> 
> Yea a few similar tests would be great

Great stuff! I'll try to get it done over next week and update the PR with some new test cases!

VikramsDataScience commented:
Hello there @peterdudfield.

I think I've completed all the remaining test cases for each of the DB endpoints for `is_fake()`. I've tested each of them to see if they're populating with data, and they appear to be working on my local machine.

Could you have a review and let me know if I've missed anything, please?

peterdudfield commented:
This looks great, do you mind putting in 2 tests for [test_national.py](https://github.com/openclimatefix/uk-pv-national-gsp-api/blob/main/src/tests/test_national.py), perhaps one testing the forecast, one testing the truth

peterdudfield commented:
It would be great to update some documentation on how to run the api on a fake data. I can imagine a docker compose file that start a database, and a api version would be good. Shall I put this in an issue, and then it could be done in a seperate PR?

peterdudfield commented on src/gsp.py:
```diff
@@ -79,10 +87,15 @@ def get_all_available_forecasts(
     logger.info(f"Get forecasts for all gsps. The option is {historic=} for user {user}")
 
-    if gsp_ids is not None:
+    if isinstance(gsp_ids, str):
```
this might have caused one of the tests to failed

VikramsDataScience replied:
Oh, that's weird. I don't have my computer on right now, but I can dig into this tomorrow, if that's cool?

peterdudfield replied:
I think it needs to happen before the `is_fake` thing

Yea no problem

VikramsDataScience replied:
I've moved the isinstance() check to occur prior to the is_fake() condition. Could trigger the docker file again to run the test, please?

VikramsDataScience replied:
I've moved the `isinstance()` check to occur prior to the `is_fake()` condition. Could you trigger the docker file again to run the test, please?

VikramsDataScience commented:
> It would be great to update some documentation on how to run the api on a fake data. I can imagine a docker compose file that start a database, and a api version would be good. Shall I put this in an issue, and then it could be done in a seperate PR?

Yeah, sure. Would you be able to assign the issue to me with as much detail as you can provide, and we can work together again to get it working, please?

VikramsDataScience commented:
> This looks great, do you mind putting in 2 tests for [test_national.py](https://github.com/openclimatefix/uk-pv-national-gsp-api/blob/main/src/tests/test_national.py), perhaps one testing the forecast, one testing the truth

Yep. No problems

VikramsDataScience commented on src/gsp.py:
```diff
@@ -251,9 +270,15 @@ def get_truths_for_all_gsps(
-    if gsp_ids is not None:
+    if isinstance(gsp_ids, str):
         gsp_ids = [int(gsp_id) for gsp_id in gsp_ids.split(",")]
 
```
I've moved the `isinstance()` check to occur prior to the `is_fake()` condition. Could trigger the docker file again to run the test, please?

peterdudfield replied:
to trigger the test, you need to close and open the PR. 

I need to put a different trigger in the workflow to make this happen automatically

VikramsDataScience commented:
> This looks great, do you mind putting in 2 tests for [test_national.py](https://github.com/openclimatefix/uk-pv-national-gsp-api/blob/main/src/tests/test_national.py), perhaps one testing the forecast, one testing the truth

I've added the 2 test cases to the test_national.py module. I'm not sure if the DB endpoints are correct or if I've even added the desired test cases. Could you review and let me know, please?

I also noticed that I had an extra for loop in my test case logic that wasn't actually needed! I've modified all the test cases to clean up the logic a little bit, as well :).

peterdudfield commented on src/tests/test_national.py:
```diff
@@ -255,3 +256,97 @@ def test_read_truth_national_gsp(db_session, api_client):
+    response = api_client.get("/v0/solar/GB/national/forecast")
+    assert response.status_code == 200
+
+    gsp_ids = [int(gsp_id) for gsp_id in range(GSP_TOTAL)]
```
to make the tests slightly simpler, I would just set `gsp_ids = [0]`

peterdudfield commented on src/gsp.py:
```diff
@@ -80,11 +88,17 @@ def get_all_available_forecasts(
 
+    if is_fake:
+        if gsp_ids is None:
+            gsp_ids = [int(gsp_id) for gsp_id in range(GSP_TOTAL)]
```
I think this should be `range(1, GSP_TOTAL)` and same with below. Its becasue (weirdly) `gsp_id=0` is the national forecast

peterdudfield commented:
> > This looks great, do you mind putting in 2 tests for [test_national.py](https://github.com/openclimatefix/uk-pv-national-gsp-api/blob/main/src/tests/test_national.py), perhaps one testing the forecast, one testing the truth
> 
> I've added the 2 test cases to the test_national.py module. I'm not sure if the DB endpoints are correct or if I've even added the desired test cases. Could you review and let me know, please?
> 
> I also noticed that I had an extra for loop in my test case logic that wasn't actually needed! I've modified all the test cases to clean up the logic a little bit, as well :).

Thanks for doing all this. 

Probably there's a few things to do
1. fix test, the comments above might help
2. add "Fake" logic into [national.py](https://github.com/openclimatefix/uk-pv-national-gsp-api/blob/main/src/national.py)

peterdudfield commented on src/tests/test_national.py:
```diff
@@ -255,3 +256,97 @@ def test_read_truth_national_gsp(db_session, api_client):
+        model_name="fake_model",
+    )
+    db_session.add_all(forecasts)
+    db_session.commit()
```
Please remove, we dont need to add these in the test, they should be add under `is_fake` in national.py

VikramsDataScience replied:
I've removed the calls to the DB in the test cases. Could you review, and let me know if I've done it correctly, please?

peterdudfield commented on src/tests/test_gsp.py:
```diff
@@ -418,3 +420,185 @@ def test_read_truths_for_all_gsp_compact(db_session, api_client):
+        model_name="fake_model",
+    )
+    db_session.add(forecasts)
+    db_session.commit()
```
remove adding forecasts in tests, this  is already done in the code you put into gsp.py

VikramsDataScience replied:
I've removed the calls to the DB in the test cases. Could you review, and let me know if I've done it correctly, please?

peterdudfield replied:
Have you? I can still see the

VikramsDataScience commented:
Ah, right. Yeah, looks we've still got some way to go, and quite a few bases yet to cover!

peterdudfield commented:
> Ah, right. Yeah, looks we've still got some way to go, and quite a few bases yet to cover!

Great progress thought. I think you've cracked the hardess part. Thanks again for doing this work

VikramsDataScience commented:
> > Ah, right. Yeah, looks we've still got some way to go, and quite a few bases yet to cover!
> 
> Great progress thought. I think you've cracked the hardess part. Thanks again for doing this work

No problems! I'm just working through some of this now. Maybe give me a few days to make the changes?

peterdudfield commented:
Yea, no worries. Take your time

peterdudfield commented on src/tests/test_gsp.py:
```diff
@@ -418,3 +419,169 @@ def test_read_truths_for_all_gsp_compact(db_session, api_client):
+        add_latest=True,
+        historic=True,
+        model_name="fake_model",
+    )
```
You need to use `make_fake_forecast` here. You should use the response object like [this](https://github.com/openclimatefix/uk-pv-national-gsp-api/blob/main/src/tests/test_national.py#L38), [here](https://github.com/openclimatefix/uk-pv-national-gsp-api/blob/main/src/tests/test_gsp.py#L41) or [here](https://github.com/openclimatefix/uk-pv-national-gsp-api/blob/main/src/tests/test_gsp.py#L115)

I think this comment goes for the other tests too.

VikramsDataScience replied:
Ah, this makes more sense. Apologies for the confusion here! I was quite unsure how to generate the response object in the tests, but this should help clear things up. I think your first option would probably be the most suitable one

peterdudfield commented:
I think its nearly there, remove those `make_fake_forecast`, and then we'll need to make sure the tests are passing. Are you able to get the tests passing locally?

VikramsDataScience commented:
> I think its nearly there, remove those `make_fake_forecast`, and then we'll need to make sure the tests are passing. Are you able to get the tests passing locally?

Yeah, they are passing locally. I'm a bit pedantic about testing my code locally before a push. Haha. 

Also, on that note, I quite foolishly left the local relative imports on the modules. Jeez, I'm very sorry about my lack of attention to detail recently 🤦‍♂️!

peterdudfield commented:
> > I think its nearly there, remove those `make_fake_forecast`, and then we'll need to make sure the tests are passing. Are you able to get the tests passing locally?
> 
> Yeah, they are passing locally. I'm a bit pedantic about testing my code locally before a push. Haha.
> 
> Also, on that note, I quite foolishly left the local relative imports on the modules. Jeez, I'm very sorry about my lack of attention to detail recently 🤦‍♂️!

Thats great you run the tests first. I wonder why they pass locally and not on CI? 

No problem, I'm very grateful you are doing this work

VikramsDataScience commented:
> > > I think its nearly there, remove those `make_fake_forecast`, and then we'll need to make sure the tests are passing. Are you able to get the tests passing locally?
> > 
> > 
> > Yeah, they are passing locally. I'm a bit pedantic about testing my code locally before a push. Haha.
> > Also, on that note, I quite foolishly left the local relative imports on the modules. Jeez, I'm very sorry about my lack of attention to detail recently 🤦‍♂️!
> 
> Thats great you run the tests first. I wonder why they pass locally and not on CI?
> 
> No problem, I'm very grateful you are doing this work

It's a really good question! I'm not too sure. On one of my earlier screenshots, those print statements that were being generated with the 0.0 and also some forecast values were what I was seeing when running them locally. So, I'm not really sure why they're failing on the CI workflows 🤔.

VikramsDataScience commented:
Apologies for the time lag in updating the PR!

I've written a simplified test case to return the `NationalForecastValue()` object, and thought I might just send a screenshot, and if you're okay with the logic, I can then do a full update to the PR?:
![image](https://github.com/user-attachments/assets/a349fa96-0023-475e-89fd-e5be0b3b9066)
And, this is the screenshot of the local tests I tend to run and it's passing locally:
![image](https://github.com/user-attachments/assets/bb07c4b3-7061-45f2-b921-be28526568da)
What do you think of the above? Should I just write simplified test cases like the above, or something different do you think?

peterdudfield commented:
Yea I think that test case is good

Thank you

VikramsDataScience commented:
> Yea I think that test case is good
> 
> Thank you

Cool. I'll make the changes as soon as I can and update the PR! 

For the GSP module, I might attach a new screenshot for you since the logic will be returning either `ForecastValue` or the `ManyForecasts` objects. Just so that we're showing that the test cases pass on a local machine 😊. I really hope that they'll pass the CI workflow, as well 🤞!

VikramsDataScience commented:
I've finished writing the test cases using the `NationalForecastValue`, `ForecastValue`, and the `ManyForecasts` as the return objects for the `test_gsp` and `test_national` modules. I've run all the test cases locally and I'm happy to report that they're are passing! If you'd like, I can provide screenshots, as well?

Otherwise, I'm hopeful that the CI workflow will pass now!

VikramsDataScience commented:
I've gone through the error log statement generated by the runner, and it looks like it's failing several of the pre-existing test cases. But, so far, the error log hasn't flagged any of the new `is_fake` test cases failing:
![image](https://github.com/user-attachments/assets/4403c050-ba6c-418b-a40e-cebeef2b33b5)
  With the above, I've gone through all the test cases that the runner has flagged in the error, and I'm able to recreate all of them locally, except for 1 - the `test_merged_routes.py::test_read_forecast_values_gsp` is passing locally:
![image](https://github.com/user-attachments/assets/78c2a4cc-9251-4356-9024-2de3dd5b1e91)

It appears that the new `is_fake` test cases are running but the 8 other test cases in the `test_gsp`, and `test_national` modules are failing.

Not sure how you'd like to approach resolving these issues, given that they're out-of-scope for this PR? Maybe to raise some new issues that the community could look into to address these failed test cases? Perhaps these test cases' logic are out of date with new changes, or something else do you think?

Also, if this means raising new issues with the community, how would you like to approach merging all the changes made in this PR? They appear to be working as expected.

peterdudfield commented:
They were all passing on 'main'. Locally do you run all the tests? 
It might be that the ENV VAR is still set?

VikramsDataScience commented:
> They were all passing on 'main'. Locally do you run all the tests? It might be that the ENV VAR is still set?

Hmm, I'm sceptical that running all the tests will yield a different result. I'm very open to any suggestions you might have, as I'm starting to run out of ideas, given how long we've been working on this PR? And, of course, you know this codebase far better than myself!

Do you mean that is_fake is being set whilst these other test cases are run?

peterdudfield commented:
> > They were all passing on 'main'. Locally do you run all the tests? It might be that the ENV VAR is still set?
> 
> Hmm, I'm sceptical that running all the tests will yield a different result. I'm very open to any suggestions you might have, as I'm starting to run out of ideas, given how long we've been working on this PR? And, of course, you know this codebase far better than myself!
> 
> Do you mean that is_fake is being set whilst these other test cases are run?

I think its worth running all the tests, to see if its the same results as CI. 

Yea, my guess is that is_fake is still set on the other cases. And in some ways needs to be un-set. 

Yea totally understand you've been working hard on this. Another options is we merge to a non `main` branch, and I, or someone else fixes the tests or other things. Its totally up to you

VikramsDataScience commented:
All good. I'm very happy to help!

I'll keep trying over next week and see how I fair ☺️

VikramsDataScience commented:
Sorry, Peter, once again a few other priorities have come up this week that need my attention. 

I did have a play around yesterday with the test cases, and I may have a couple of theories as to why they're failing, but haven't experimented to see my theories pan out! Once I get these other priorities out of the way, I'll return to the PR :).

VikramsDataScience commented:
Hey Peter. So with these test cases, I've just started looking into them. And, as you correctly suggested, to run them all together, rather than individually. Locally, I'm able to reproduce 8/9 failures in the CI workflow. 

And, looking at the individual test cases, I was able to 'fix' two of them - `test_read_latest_all_gsp()` and `test_read_latest_all_gsp_normalized()` by adding one extra 'gsp_id' to their DB call (by changing `gsp_ids=list(range(0, 11))`), since the failure was caused by the first assert statement that's testing for the presence of 10 forecasts in the object, but 9 were returning. 

The other test cases have different failure conditions, at least 3 of them seem to be caused by duplicate values in the DB that are persisting from the previous test cases that added those values:
![image](https://github.com/user-attachments/assets/5ef7d107-ba51-472a-a3e2-ce077a861717)

The question (and permission, really) I wanted to ask is that can I alter the conditions of these cases to stop them from failing our new merge? But my fear is that, based on your previous comment, you were saying they're currently passing on the main branch? So, if I alter them to pass on this new merge, I'm afraid that it would damage the integrity of the existing logic, and then they might begin failing afterward? Or that the new logic might end up testing the wrong thing! If you're cool with me altering the code, I'm happy to keep working on it until they pass, but I just don't want to damage integrity of the logic! Please let me know how you'd like to proceed :).

VikramsDataScience commented:
Hey Peter. So with these test cases, I've just started looking into them. And, as you correctly suggested, to run them all together, rather than individually. Locally, I'm able to reproduce 8/9 failures in the CI workflow (on my local, the test_read_forecast_values_gsp()) is working.

And, looking at the individual test cases, I was able to 'fix' two of them - test_read_latest_all_gsp() and test_read_latest_all_gsp_normalized() by adding one extra 'gsp_id' to their DB call (by changing gsp_ids=list(range(0, 11))), since the failure was caused by the first assert statement that's testing for the presence of 10 forecasts in the return object, but 9 were returning.

The other test cases have different failure conditions - they seem to be caused by duplicate values in the DB that are persisting from the previous test cases that added those values:
![image](https://github.com/user-attachments/assets/4afcf4cb-63d7-4e88-9c48-9780545684bc)

The question (and permission, really) I wanted to ask is that can I alter the conditions of these cases to stop them from failing our new merge? But my fear is that, based on your previous comment, you were saying they're currently passing on the main branch? So, if I alter them to pass on this new merge, I'm afraid that it would damage the integrity of the existing logic, and then they might begin failing afterward? Or that the new logic might end up testing the wrong thing! If you're cool with me altering the code, I'm happy to keep working on it until they pass, but I just don't want to damage integrity of the logic! Please let me know how you'd like me to proceed :).

VikramsDataScience commented:
Small update and a bit of good news to the previous comment:
![image](https://github.com/user-attachments/assets/d9f8989c-e01e-400d-b027-298a7ed00cd2)
I've managed to fix 3 of the test cases by applying a `pytest.fixture()` to the `setup_gsp_yield_data()` the helper function, and then yielding the `db_session` in the test cases that call it: `test_read_truths_for_all_gsp()`, `test_read_truths_for_all_gsp()`, and `test_read_truths_for_all_gsp_filter_gsp()`. The good new is that this was done without changing any of the underlying logic in those test cases. But, please do let me know about how you'd like me to proceed with the points in the previous comment.

peterdudfield commented:
> Small update and a bit of good news to the previous comment: ![image](https://private-user-images.githubusercontent.com/45002417/387058477-d9f8989c-e01e-400d-b027-298a7ed00cd2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE5MTk5OTIsIm5iZiI6MTczMTkxOTY5MiwicGF0aCI6Ii80NTAwMjQxNy8zODcwNTg0NzctZDlmODk4OWMtZTAxZS00MDBkLWIwMjctMjk4YTdlZDAwY2QyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE4VDA4NDgxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM4ZTU3NDVhYjcyNWQ0NTU1MGRhYzVmZTkzM2U4ZDBlMWExZTBmOGQ5NjQzMmZkOTNjZTBhZDhlZmE5YjgxMzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WTie_uplzQRzZQwzwRlYuOIhLBbHtDQNM-9oQSgd2XM) I've managed to fix 3 of the test cases by applying a `pytest.fixture()` to the `setup_gsp_yield_data()` the helper function, and then yielding the `db_session` in the test cases that call it: `test_read_truths_for_all_gsp()`, `test_read_truths_for_all_gsp()`, and `test_read_truths_for_all_gsp_filter_gsp()`. The good new is that this was done without changing any of the underlying logic in those test cases. But, please do let me know about how you'd like me to proceed with the points in the previous comment.

Thats great, good to do that. Can you push that? Then we can see what other tests are failing

VikramsDataScience commented:
I've made the changes from the above comments, and I'm happy to say that we've now made some good progress in solving our issues. So, we should have fewer errors in the next workflow run. These are some of the changes:
- Explicitly disabled `is_fake()` in the new test cases to prevent any issues that may come up that the is_fake ENV may cause.
- Fixed the `test_gsp` issues from the previous comments by individually yielding the stored values in the db_session to prevent the duplication errors we've been seeing.
- Also applied the same fix for the `test_national` `test_read_truth_national_gsp()` test cases, as the error from the CI workflow seems to come from the same source.
- I've also rolled back the change I made to address that assertion issue from `test_gsp.py::test_read_latest_all_gsp` and `test_gsp.py::test_read_latest_all_gsp_normalized`. So, we'll continue to get those errors in the CI workflow. Maybe these two test cases are not including the national GSP for some reason (`gsp_id==0`)? From what I could see when I ran an assert on the `forecasts` object, there were 10 forecasts in that object, but when these forecasts are inserted to the DB, the gsp_id==0 gets omitted for some weird reason, I think. Keen to hear any thoughts you might have on how we can solve for this issue?:
![image](https://github.com/user-attachments/assets/c2e74cdb-19a6-4149-a66a-fe9095ff45eb)
- I'm not able to locally reproduce the error from the `test_merged_routes.py::test_read_forecast_values_gsp` that the workflow is generating (it's passing on my machine). So I'm not sure how to repair that error when it comes up again. Very open to any suggestions for that one, as well?

If my local tests work as expected in the workflow, I'd reckon that we will probably get 4 errors, rather than 9, in the next run.

VikramsDataScience commented:
Hey Peter,

I may have a fix for the `test_read_latest_all_gsp_normalized()` and `test_read_latest_all_gsp()` test cases. It took me quite some time to try to debug this issue! I had to write and test a variety of statements to see where the `gsp_id==0` was dropping off, but after quite a bit of testing and experimentation, I tried using the 'gsp_ids' as a query parameter in the DB endpoint, and now the `test_gsp.py` passes all the tests on my local machine:
![image](https://github.com/user-attachments/assets/1d5979ad-c088-46ef-990a-5e4bedee946c)
This should now mean 2 remaining failed cases:
- `test_merged_routes.py::test_read_forecast_values_gsp`
- `test_national.py::test_read_truth_national_gsp`

VikramsDataScience commented:
- I think that our remaining errors may be to do with incorrect DB endpoints.
- On that note, I'm trying an experiment to address the `test_read_truth_national_gsp()` and `test_read_forecast_values_gsp()` failures. Unfortunately, for the remaining two test cases, I can't reproduce them on my local machine for some weird reason. So I'll only be able to test any changes by pushing to the PR and examining the error logs generated by the runners. So, you may get a few pushes to the PR as I try work through this!

VikramsDataScience commented:
Hey Peter,

Are you able to trigger the workflow to see my newest changes have any effect?

peterdudfield commented:
I think you can close and reop the PR, adn that triggers them


Peter Dudfield
+44 7814 169 251
________________________________
From: Vikram Pande ***@***.***>
Sent: 22 November 2024 09:47
To: openclimatefix/uk-pv-national-gsp-api ***@***.***>
Cc: Peter Dudfield ***@***.***>; State change ***@***.***>
Subject: Re: [openclimatefix/uk-pv-national-gsp-api] Fake data Issue #354 (PR #360)


Hey Peter,

Are you able to trigger the workflow to see my newest changes have any effect?

—
Reply to this email directly, view it on GitHub<https://github.com/openclimatefix/uk-pv-national-gsp-api/pull/360#issuecomment-2493349016>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AIIUKWXJLBDHDEX44K65MID2B34SNAVCNFSM6AAAAABOXLO6XCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIOJTGM2DSMBRGY>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>

VikramsDataScience commented:
> I think you can close and reop the PR, adn that triggers them Peter Dudfield +44 7814 169 251
> […](#)
> ________________________________ From: Vikram Pande ***@***.***> Sent: 22 November 2024 09:47 To: openclimatefix/uk-pv-national-gsp-api ***@***.***> Cc: Peter Dudfield ***@***.***>; State change ***@***.***> Subject: Re: [openclimatefix/uk-pv-national-gsp-api] Fake data Issue #354 (PR #360) Hey Peter, Are you able to trigger the workflow to see my newest changes have any effect? — Reply to this email directly, view it on GitHub<[#360 (comment)](https://github.com/openclimatefix/uk-pv-national-gsp-api/pull/360#issuecomment-2493349016)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AIIUKWXJLBDHDEX44K65MID2B34SNAVCNFSM6AAAAABOXLO6XCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIOJTGM2DSMBRGY>. You are receiving this because you modified the open/close state.Message ID: ***@***.***>

No problems. I wasn't sure if I had sufficient permissions, but I can do that from now :)

VikramsDataScience commented:
Whoa! I think it may have worked!

peterdudfield commented:
Thats great they are now working. I relised something when reviewing this. For the routes that say `get the truth` we need to not use `make_fake_forecast` but use make [fake gsp yeilds](https://github.com/openclimatefix/nowcasting_datamodel/blob/main/nowcasting_datamodel/fake.py#L264). Would you be able to change this over? I suspect we might be able to role back some changes and simplify this PR then. 

Again, thanks so much for this work, and its very nearly over the line. I'd be very happy if you do this bit, but I totally understand if you want me or someone else to do it

VikramsDataScience commented:
That's cool. I should be able to take care of it and update the PR if all is well, and otherwise comment if I have any challenges :)!

VikramsDataScience commented:
Hey Peter,

Apologies for going a bit quiet. I've grown busy once again over the last week, or so. I should get a chance to look into this sometime this week. Just thought I'd provide an update :)

peterdudfield commented:
no problem, thank you

VikramsDataScience commented:
Hey Peter,

I'm just looking into this one now. With the call to the `make_fake_gsp_yields()` function, it seems that I'd need to make calls to the DB in the test cases? Something like the below screenshot:
![image](https://github.com/user-attachments/assets/adfebf95-31f4-42d1-8aed-b1a7f2c62d2a)
In some of my earlier commits, I remember you were saying not make calls to the DB in the test cases, since the `is_fake()` ENV is already doing that in the `gsp.py` and `national.py` modules? Please let me know if something resembling the screenshot is okay, or should I be doing something different?

peterdudfield commented:
> make_fake_gsp_yields

ah, yea. You shouldnt need to do this in the test cases. It would be good if you put this in that actual routes. So when pulling `get_national_pvlive` you should use `make_fake_gsp_yields` not `make_fake_forecast`. These should get added to the database already, so you shouldnt have to use session.add e.t.c

VikramsDataScience commented:
Hey Peter,

I've experimented a bit with the `make_fake_gsp_yields()` function in this test case. So far, I've only set it to test for the presence of a returned object:
![image](https://github.com/user-attachments/assets/af9753cc-1817-470e-b641-860c205c6aae)
However, the `make_fake_gsp_yields()` and the corresponding `make_fake_gsp_yields_for_one_location()` aren't returning any objects once the values are added to the DB. Should I be constructing the test case differently, since they're not returning any objects?

peterdudfield commented:
> Hey Peter,
> 
> I've experimented a bit with the `make_fake_gsp_yields()` function in this test case. So far, I've only set it to test for the presence of a returned object: ![image](https://private-user-images.githubusercontent.com/45002417/394069415-af9753cc-1817-470e-b641-860c205c6aae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4MTk0NTYsIm5iZiI6MTczMzgxOTE1NiwicGF0aCI6Ii80NTAwMjQxNy8zOTQwNjk0MTUtYWY5NzUzY2MtMTgxNy00NzBlLWI2NDEtODYwYzIwNWM2YWFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDA4MjU1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEzZjM5YTQwYjJkMzA3YzAzNTQzN2FiYTg2YjY4YmNlMWY5YWRkYmJkNDdiYjljYmI0MDE0ZjI4NWRhMmFjNmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qrHgCb5fW8gnKxqti0td3sUobXzKVm6Ilfo9g2VbfwI) However, the `make_fake_gsp_yields()` and the corresponding `make_fake_gsp_yields_for_one_location()` aren't returning any objects once the values are added to the DB. Should I be constructing the test case differently, since they're not returning any objects?

Hey, 

yea, you don't need to add these to the test cases. 
You should use them in the api routes.

peterdudfield commented on src/national.py:
```diff
@@ -204,8 +214,11 @@ def get_national_pvlive(
     logger.info(f"Get national PV Live estimates values " f"for regime {regime} for  {user}")
 
+    if is_fake:
+        make_fake_forecast(gsp_id=gsp_id, session=session)
```
This is where we should use make_fake_gsp_yields, and not make_fake_forecast

peterdudfield commented on src/gsp.py:
```diff
@@ -286,6 +313,10 @@ def get_truths_for_a_specific_gsp_old_route(
     """Redirects old API route to new route /v0/solar/GB/gsp/{gsp_id}/pvlive"""
+
+    if is_fake:
+        make_fake_forecast(gsp_id=gsp_id, session=session)
```
This is where we should use `make_fake_gsp_yields`, and not `make_fake_forecast`

VikramsDataScience replied:
Ah, right. That makes more sense! I've made and pushed the changes now. I might need to run a few unsuccessful tests before getting it right :).

peterdudfield commented on src/gsp.py:
```diff
@@ -251,11 +271,18 @@ def get_truths_for_all_gsps(
+        if gsp_ids is None:
+            gsp_ids = [int(gsp_id) for gsp_id in range(1, GSP_TOTAL)]
+
+        make_fake_forecasts(gsp_ids=gsp_ids, session=session)
```
This is where we should use `make_fake_gsp_yields`, and not `make_fake_forecast`

VikramsDataScience replied:
Ah, right. That makes more sense! I've made and pushed the changes now. I might need to run a few unsuccessful tests before getting it right :).

VikramsDataScience commented:
Hey Peter,

Looks like our latest errors are still to do with how the test cases test the DB routes. I've fixed 2 of the 3 errors in the CI workflow that were to do with the `test_gsp` test cases, but I'm still having problems with the `test_national` test cases. Here's a quick summary:

With the `test_gsp` test cases:
- I've used the `LocationWithGSPYields()` class as the return object in the `test_is_fake_get_truths_for_all_gsps()` as this seems to be the object that's returned in the `get_truths_for_all_gsps()` function. Not sure if that's correct?

With the `test_national` test cases:
- I'm still having issues with the `/national/pvlive` route. I've tried numerous combinations, but to no avail. Not sure if you have any feedback for the correct route to use in the `test_is_fake_national_get_truths_for_all_gsps()` that will resolve the final failed test case? I think once we resolve this final failed test case, we should be good to go.

peterdudfield commented on src/tests/test_national.py:
```diff
@@ -248,10 +249,47 @@ def test_read_truth_national_gsp(db_session, api_client):
     _ = [GSPYield(**gsp_yield) for gsp_yield in r_json]
+
+
+def test_is_fake_national_all_available_forecasts(monkeypatch, api_client):
```
nice, these tests are in principle corretc, thank you for putting them in

VikramsDataScience replied:
Awesome! Thanks for this. Looks like it's worked!

peterdudfield commented on src/gsp.py:
```diff
@@ -286,6 +313,10 @@ def get_truths_for_a_specific_gsp_old_route(
     """Redirects old API route to new route /v0/solar/GB/gsp/{gsp_id}/pvlive"""
+
+    if is_fake:
+        make_fake_gsp_yields(gsp_ids=List[gsp_id], session=session)
```
should this just be `[gsp_id]` not `List[gsp_ids]`

VikramsDataScience replied:
Awesome! Thanks for this. Looks like it's worked!

peterdudfield commented on src/gsp.py:
```diff
@@ -251,11 +271,18 @@ def get_truths_for_all_gsps(
 
+    if is_fake:
+        if gsp_ids is None:
+            gsp_ids = [List[int(gsp_id)] for gsp_id in range(1, GSP_TOTAL)]
```
should this be `gsp_ids = [int(gsp_id) for gsp_id in range(1, GSP_TOTAL)]`

VikramsDataScience replied:
Awesome! Thanks for this. Looks like it's worked!

peterdudfield commented on src/national.py:
```diff
@@ -204,8 +214,11 @@ def get_national_pvlive(
     logger.info(f"Get national PV Live estimates values " f"for regime {regime} for  {user}")
 
+    if is_fake:
+        make_fake_gsp_yields(gsp_ids=List[gsp_id], session=session)
```
should this be `make_fake_gsp_yields(gsp_ids=[gsp_id], session=session)`

VikramsDataScience replied:
Awesome! Thanks for this. Looks like it's worked!

peterdudfield commented:
> Hey Peter,
> 
> Looks like our latest errors are still to do with how the test cases test the DB routes. I've fixed 2 of the 3 errors in the CI workflow that were to do with the `test_gsp` test cases, but I'm still having problems with the `test_national` test cases. Here's a quick summary:
> 
> With the `test_gsp` test cases:
> 
> * I've used the `LocationWithGSPYields()` class as the return object in the `test_is_fake_get_truths_for_all_gsps()` as this seems to be the object that's returned in the `get_truths_for_all_gsps()` function. Not sure if that's correct?
> 
> With the `test_national` test cases:
> 
> * I'm still having issues with the `/national/pvlive` route. I've tried numerous combinations, but to no avail. Not sure if you have any feedback for the correct route to use in the `test_is_fake_national_get_truths_for_all_gsps()` that will resolve the final failed test case? I think once we resolve this final failed test case, we should be good to go.

Thanks for all this. Ive made a few tiny comments that might solve it

VikramsDataScience commented:
Thank you for the feedback, Peter! Looks like the CI workflow has now passed. I think we've finally arrived 🎉!

peterdudfield reviewed (approved):
Well done with all this. Thank you so muhc for doing this. 

There's a tiny change ive written, but i think then good to go

VikramsDataScience commented:
> Well done with all this. Thank you so muhc for doing this.
> 
> There's a tiny change ive written, but i think then good to go

No problems! I'll make the change tomorrow and make another push.

Are there any other issues that I can work on after this that are of similar or greater complexity than this one that you'd like me to work on? Especially once the NY begins?

peterdudfield commented:
> > Well done with all this. Thank you so muhc for doing this.
> > There's a tiny change ive written, but i think then good to go
> 
> No problems! I'll make the change tomorrow and make another push.
> 
> Are there any other issues that I can work on after this that are of similar or greater complexity than this one that you'd like me to work on? Especially once the NY begins?

Thanks, yea good first issue are alwasy good. A good follow on from this one, is to check the quartz solar UI works with the Fake API, perhaps that could be something you tackle

VikramsDataScience commented:
> > > Well done with all this. Thank you so muhc for doing this.
> > > There's a tiny change ive written, but i think then good to go
> > 
> > 
> > No problems! I'll make the change tomorrow and make another push.
> > Are there any other issues that I can work on after this that are of similar or greater complexity than this one that you'd like me to work on? Especially once the NY begins?
> 
> Thanks, yea good first issue are alwasy good. A good follow on from this one, is to check the quartz solar UI works with the Fake API, perhaps that could be something you tackle. I can write that up in more detail if you like

Yes, please. If possible could you assign to me, and I can start looking into it. Most likely in the NY if that's cool with you?

peterdudfield commented:
here we go @VikramsDataScience https://github.com/openclimatefix/uk-pv-national-gsp-api/issues/367

VikramsDataScience commented:
Actually they were placed there to fix the previously failed test cases. If we remove them those test cases will likely fail again. This was my comment and your reply about a month ago (https://github.com/openclimatefix/uk-pv-national-gsp-api/pull/360#issuecomment-2482306669)

peterdudfield commented:
> Actually they were placed there to fix the previously failed test cases. If we remove them those test cases will likely fail again. This was my comment and your reply about a month ago ([#360 (comment)](https://github.com/openclimatefix/uk-pv-national-gsp-api/pull/360#issuecomment-2482306669))

Hey, ok, they shoudlnt be needed though,  so worth trying to remove and seeing if they work.

VikramsDataScience commented:
Without yielding values from the DB, and without the `fixture` capability the test cases are now failing again. Should I revert my original solution with the fixture and DB yields?

VikramsDataScience commented on src/tests/test_gsp.py:
```diff
@@ -405,6 +420,7 @@ def test_read_truths_for_all_gsp_compact(db_session, api_client):
     setup_gsp_yield_data(db_session=db_session)
 
     app.dependency_overrides[get_session] = lambda: db_session
+    yield db_session
```
Without yielding values from the DB, and without the `fixture` capability the test cases are now failing. Should I revert my original solution with the fixture and DB yields?

peterdudfield commented:
Looking at the logs for the test, it looks like some of the non fake tests are failing. I suspect this is due to FAKE still being set. This might be becasue some tests are run in parrellel. The yield int he middle of the tests, i casuing ti to be skipped. 

One solution cold be to make a separate folder call tests/fake/ and put all the fake tests in there. Maybe this is worht ago?

VikramsDataScience commented:
Yeah, I think all the failing test cases are NON-FAKE. I suspect if we split my new FAKE test cases into a separate test module these will probably work again. 

One question, though (might be a dumb question) - if I put the FAKE test cases into a separate sub-directory `/tests/fake`, will the `pytest` command from the CI workflow still execute my new FAKE test cases?

peterdudfield commented:
> Yeah, I think all the failing test cases are NON-FAKE. I suspect if we split my new FAKE test cases into a separate test module these will probably work again.
> 
> One question, though (might be a dumb question) - if I put the FAKE test cases into a separate sub-directory `/tests/fake`, will the `pytest` command from the CI workflow still execute my new FAKE test cases?

Yea it should run them, and it normally runs a file at a time

VikramsDataScience commented:
That's good. I'll make the change tomorrow and push. Hopefully it'll work. 

Will you be off for a couple of weeks?

peterdudfield commented:
> That's good. I'll make the change tomorrow and push. Hopefully it'll work.
> 
> Will you be off for a couple of weeks?

I off for holidays, but Im sure I can have a look at any changes you make

VikramsDataScience commented:
All good. I'm looking to take some time off as well. Unless I get bored, of course!

VikramsDataScience commented:
Hey Peter,
- I've made the changes to the `is_fake` test cases in the `test_gsp` module by creating a new module called `/tests/fake/test_gsp_fake.py`. 
However, without the DB yields and pytest fixture solution, the test cases are still failing the CI workflow with the same errors, but the new is_fake test module does work on its own. 
My question here is even without the presence of the is_fake test cases these other test cases are still failing the CI workflow, how are they passing in the main branch? I mean, other than my proposed solution using pytest fixture and the resulting DB yields, I haven't changed the underlying logic for the other test cases, so why would they fail here, do you think? It's quite strange.
- For now, I've left the national `is_fake` test cases unchanged in the `test_national` module since is passing, but I can create a new `/tests/fake/test_national_fake.py` if we find a solution that works.

peterdudfield commented:
> Hey Peter,
> 
> * I've made the changes to the `is_fake` test cases in the `test_gsp` module by creating a new module called `/tests/fake/test_gsp_fake.py`.
>   However, without the DB yields and pytest fixture solution, the test cases are still failing the CI workflow with the same errors, but the new is_fake test module does work on its own.
>   My question here is even without the presence of the is_fake test cases these other test cases are still failing the CI workflow, how are they passing in the main branch? I mean, other than my proposed solution using pytest fixture and the resulting DB yields, I haven't changed the underlying logic for the other test cases, so why would they fail here, do you think? It's quite strange.
> * For now, I've left the national `is_fake` test cases unchanged in the `test_national` module since is passing, but I can create a new `/tests/fake/test_national_fake.py` if we find a solution that works.

Shall we merge this to a branch called fake-data in the repo, then I could have a go at the tests?

VikramsDataScience commented:
> > Hey Peter,
> > 
> > * I've made the changes to the `is_fake` test cases in the `test_gsp` module by creating a new module called `/tests/fake/test_gsp_fake.py`.
> >   However, without the DB yields and pytest fixture solution, the test cases are still failing the CI workflow with the same errors, but the new is_fake test module does work on its own.
> >   My question here is even without the presence of the is_fake test cases these other test cases are still failing the CI workflow, how are they passing in the main branch? I mean, other than my proposed solution using pytest fixture and the resulting DB yields, I haven't changed the underlying logic for the other test cases, so why would they fail here, do you think? It's quite strange.
> > * For now, I've left the national `is_fake` test cases unchanged in the `test_national` module since is passing, but I can create a new `/tests/fake/test_national_fake.py` if we find a solution that works.
> 
> Shall we merge this to a branch called fake-data in the repo, then I could have a go at the tests?

Sure. Would you be okay to tag me when you are able to repair this issue, since the next issue would be dependent on this one working, please?
